// nativescript-dev-appium/lib/appium-server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process = require("child_process");
const utils_1 = require("./utils");
const device_controller_1 = require("./device-controller");
const mobile_devices_controller_1 = require("mobile-devices-controller");
class AppiumServer {
    constructor(_args) {
        this._args = _args;
        this._runType = this._args.runType;
        this._hasStarted = false;
        this.resolveAppiumDependency();
    }
    get port() {
        return this._port;
    }
    set port(port) {
        this._port = port;
    }
    set runType(runType) {
        this._runType = runType;
    }
    get runType() {
        return this._runType;
    }
    get server() {
        return this._server;
    }
    get hasStarted() {
        return this._hasStarted;
    }
    set hasStarted(hasStarted) {
        this._hasStarted = hasStarted;
    }
    start(port, deviceManager = new device_controller_1.DeviceManger()) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.prepareDevice(deviceManager);
            yield this.prepareApp();
            utils_1.log("Starting server...", this._args.verbose);
            const logLevel = this._args.verbose === true ? "debug" : "info";
            this.port = this._args.port || port;
            let retry = false;
            this.startAppiumServer(logLevel);
            let response = yield utils_1.waitForOutput(this._server, /listener started/, /Error: listen/, 60000, this._args.verbose);
            let retries = 11;
            while (retries > 0 && !response) {
                retries--;
                this.port += 10;
                this.port = (yield utils_1.findFreePort(100, this.port));
                this.startAppiumServer(logLevel);
                response = yield utils_1.waitForOutput(this._server, /listener started/, /Error: listen/, 60000, true);
            }
            return response;
        });
    }
    startAppiumServer(logLevel) {
        this._server = child_process.spawn(this._appium, ["-p", this.port.toString(), "--log-level", logLevel], {
            shell: true,
            detached: false
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._deviceManager.stopDevice(this._args);
            return new Promise((resolve, reject) => {
                this._server.on("close", (code, signal) => {
                    utils_1.log(`Appium terminated due signal: ${signal} and code: ${code}`, this._args.verbose);
                    resolve();
                });
                this._server.on("exit", (code, signal) => {
                    utils_1.log(`Appium terminated due signal: ${signal} and code: ${code}`, this._args.verbose);
                    resolve();
                });
                this._server.on("error", (code, signal) => {
                    utils_1.log(`Appium terminated due signal: ${signal} and code: ${code}`, this._args.verbose);
                    resolve();
                });
                this._server.on("disconnect", (code, signal) => {
                    utils_1.log(`Appium terminated due signal: ${signal} and code: ${code}`, this._args.verbose);
                    resolve();
                });
                utils_1.log("Stopping server...", this._args.verbose);
                try {
                    if (utils_1.isWin()) {
                        utils_1.shutdown(this._server, this._args.verbose);
                        this._server.kill("SIGINT");
                        this._server.kill("SIGINT");
                        this._server = null;
                    }
                    else {
                        this._server.kill("SIGINT");
                        this._server.kill("SIGINT");
                        this._server.kill("SIGKILL");
                        utils_1.shutdown(this._server, this._args.verbose);
                    }
                }
                catch (error) {
                    console.log(error);
                }
            });
        });
    }
    prepareDevice(deviceManager) {
        return __awaiter(this, void 0, void 0, function* () {
            this._deviceManager = deviceManager;
            if (!this._args.device) {
                const device = yield this._deviceManager.startDevice(this._args);
                this._args.device = device;
            }
        });
    }
    prepareApp() {
        return __awaiter(this, void 0, void 0, function* () {
            const appPackage = this._args.isAndroid ? "appPackage" : "bundleId";
            const appFullPath = this._args.appiumCaps.app;
            if (appFullPath && !this._args.appiumCaps[appPackage]) {
                console.log(`Trying to resolve automatically ${appPackage}!`);
                this._args.appiumCaps[appPackage] = this._deviceManager.getPackageId(this._args.device, appFullPath);
                console.log(`Setting capabilities ${this._args.runType}{ "${appPackage}" : "${this._args.appiumCaps[appPackage]}" }!`);
            }
            const appActivityProp = "appActivity";
            if (this._args.isAndroid && appFullPath && !this._args.appiumCaps[appActivityProp]) {
                console.log(`Trying to resolve automatically ${appActivityProp}!`);
                this._args.appiumCaps[appActivityProp] = mobile_devices_controller_1.AndroidController.getLaunchableActivity(appFullPath);
                console.log(`Setting capabilities ${this._args.runType}{ "${appActivityProp} : "${this._args.appiumCaps[appActivityProp]}" }!`);
            }
            if (!this._args.appiumCaps[appPackage]) {
                throw new Error(`Please, provide ${appPackage} in ${this._args.appiumCapsLocation} file!`);
            }
            if (this._args.isAndroid && !this._args.appiumCaps[appActivityProp]) {
                throw new Error(`Please, provide ${appActivityProp} in ${this._args.appiumCapsLocation} file!`);
            }
            const groupings = utils_1.getRegexResultsAsArray(/(\w+)/gi, this._args.appiumCaps[appPackage]);
            this._args.appName = groupings[groupings.length - 1];
            console.log(`Setting application name as ${this._args.appName}`);
            if (!this._args.devMode) {
                yield this._deviceManager.uninstallApp(this._args);
            }
            else {
                this._args.appiumCaps.app = "";
            }
        });
    }
    // Resolve appium dependency
    resolveAppiumDependency() {
        const projectDir = this._args.projectDir;
        const pluginBinary = this._args.pluginBinary;
        const projectBinary = this._args.projectBinary;
        const pluginRoot = this._args.pluginRoot;
        let appium = process.platform === "win32" ? "appium.cmd" : "appium";
        const pluginAppiumBinary = utils_1.resolve(pluginBinary, appium);
        const projectAppiumBinary = utils_1.resolve(projectBinary, appium);
        if (utils_1.fileExists(pluginAppiumBinary)) {
            utils_1.log("Using plugin-local Appium binary.", this._args.verbose);
            appium = pluginAppiumBinary;
        }
        else if (utils_1.fileExists(projectAppiumBinary)) {
            utils_1.log("Using project-local Appium binary.", this._args.verbose);
            appium = projectAppiumBinary;
        }
        else {
            //const result = executeCommand("npm list -g");
            //if (result.includes("appium")) {
            utils_1.log("Using global Appium binary.", true);
            utils_1.log('Please, make sure it is installed globally.', true);
            //} else if (result.includes("appium")) {
            //   const msg = "Appium not found. Please install appium before runnig tests!";
            //     log(msg, this._args.verbose);
            //     new Error(msg);
            // }
        }
        this._appium = appium;
    }
}
exports.AppiumServer = AppiumServer;
